cpp/src/parquet/properties_driven_crypto_factory.cc

```diff
@@ -0,0 +1,278 @@
+    throw ParquetException(
+        "Cannot call both column_keys(const std::string&) and uniform_encryption()");
+  }
+  column_keys_ = column_keys;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::uniform_encryption() {
+  if (!column_keys_.empty()) {
+    throw ParquetException(
+        "Cannot call both column_keys(const std::string&) and uniform_encryption()");
+  }
+  uniform_encryption_ = true;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::encryption_algorithm(
+    ParquetCipher::type algo) {
+  encryption_algorithm_ = algo;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::plaintext_footer(
+    bool plaintext_footer) {
+  plaintext_footer_ = plaintext_footer;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::double_wrapping(
+    bool double_wrapping) {
+  double_wrapping_ = double_wrapping;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::wrap_locally(
+    bool wrap_locally) {
+  wrap_locally_ = wrap_locally;
+  return this;
+}
+
+EncryptionConfiguration::Builder*
+EncryptionConfiguration::Builder::cache_lifetime_seconds(
+    uint64_t cache_lifetime_seconds) {
+  cache_lifetime_seconds_ = cache_lifetime_seconds;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::internal_key_material(
+    bool internal_key_material) {
+  internal_key_material_ = internal_key_material;
+  return this;
+}
+
+EncryptionConfiguration::Builder* EncryptionConfiguration::Builder::data_key_length_bits(
+    int32_t data_key_length_bits) {
+  data_key_length_bits_ = data_key_length_bits;
+  return this;
+}
+
+std::shared_ptr<EncryptionConfiguration> EncryptionConfiguration::Builder::build() {
+  if (!uniform_encryption_ && column_keys_.empty()) {
+    throw ParquetException(
+        "Either column_keys(const std::string&) or uniform_encryption() must be "
+        "called.");
+  }
+
+  return std::make_shared<EncryptionConfiguration>(
+      footer_key_, column_keys_, encryption_algorithm_, plaintext_footer_,
+      double_wrapping_, wrap_locally_, cache_lifetime_seconds_, internal_key_material_,
+      uniform_encryption_, data_key_length_bits_);
+}
+
+DecryptionConfiguration::Builder* DecryptionConfiguration::Builder::wrap_locally(
+    bool wrap_locally) {
+  wrap_locally_ = wrap_locally;
+  return this;
+}
+
+DecryptionConfiguration::Builder*
+DecryptionConfiguration::Builder::cache_lifetime_seconds(
+    uint64_t cache_lifetime_seconds) {
+  cache_lifetime_seconds_ = cache_lifetime_seconds;
+  return this;
+}
+
+std::shared_ptr<DecryptionConfiguration> DecryptionConfiguration::Builder::build() {
+  return std::make_shared<DecryptionConfiguration>(wrap_locally_,
+                                                   cache_lifetime_seconds_);
+}
+
+void PropertiesDrivenCryptoFactory::kms_client_factory(
```

Comment:
I think it's better called "register_kms_client_factory" or "set_kms_client_factory". Also, can be made to throw an exception if already set, so a user won't override it by mistake.